clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -162,6 +163,13 @@ public void connect(String id, InetSocketAddress address, int sendBufferSize, in
         KafkaChannel channel = channelBuilder.buildChannel(id, key, maxReceiveSize);
         key.attach(channel);
         this.channels.put(id, channel);
+
+        if (connected) {
```

Comment:
Yes, that may not be enough since we need to propagate the connected state through Selector.poll(). Could you try the following?

In Selector, we maintain
Set<String> immediatelyConnected;

In connect(), we make the following changes.
If connected is true {
  immediatelyConnected.add(id);
  key = socketChannel.register(nioSelector, 0);
} else {
  key = socketChannel.register(nioSelector, SelectionKey.OP_CONNECT);
}

In poll(), we make the following changes.
1. if (hasStagedReceives()) || immedidatelyConnected.size() > 0) timeout = 0
This will make sure that select() will return immediately.
2. After the call to addToCompletedReceives(), we will add the following code.
for each id in immedidatelyConnected
  get the KafkaChannel from id
  kafkaChannel.finishConnect()
  this.connected.add(id)
immedidatelyConnected.clear();